pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -117,11 +211,105 @@ func (a Adapter) Create() (err error) {
 		return err
 	}
 
-	podSelector := fmt.Sprintf("component=%s", componentName)
+	return nil
+}
+
+// Push syncs source code from the user's disk to the component
+func (a Adapter) pushLocal(path string, files []string, delFiles []string, isForcePush bool, globExps []string) error {
```

Comment:
this function is doing alot of things, lets try to extract certain portions to make unit testing better. 
two suggestions would be be -
1 - https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R254-R264 can be extracted into a function
2 - https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R268-R277 can be extracted into a function

Also if the function has more then 5 args please try to create a struct to pass on data or consider moving it as part of the method's struct if it doesn't change through out the lifecycle of the struct. 
e.g. path and ignoredFiles fields might not change throughout the life of `Adapter` so maybe it can be places as part of the Adapter's struct? - this is just an example.

Reply:
@girishramnani I'm not sure moving those sections into separate functions would make unit testing easier - they both involve exec'ing into an already running container and manipulating the files that exist there, and that can't be mocked with the fake kube client.

Reply:
we can mock it without the fake kube client and it will definitely make unit testing pushLocal easier because when you would write test for pushLocal you would mock the https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R254-R264 and https://github.com/openshift/odo/pull/2681/files#diff-825a1a88d75e789899c58e89a79eb767R268-R277 functions.

Reply:
I think what makes it different is that we're interacting with the kube client different than  we do for other resources, we're interacting directly with the rest client, not calling an existing `exec` function:
```
	req := c.KubeClient.CoreV1().RESTClient().
		Post().
		Namespace(c.Namespace).
		Resource("pods").
		Name(podName).
		SubResource("exec").
		VersionedParams(&podExecOptions, scheme.ParameterCodec)
```
and
```
	exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
	if err != nil {
		return errors.Wrapf(err, "unable execute command via SPDY")
	}
	// initialize the transport of the standard shell streams
	err = exec.Stream(remotecommand.StreamOptions{
		Stdin:  stdin,
		Stdout: stdout,
		Stderr: stderr,
		Tty:    tty,
	})
```

(see https://github.com/openshift/odo/blob/master/pkg/kclient/pods.go#L93)

Reply:
As a compromise, rather than putting those sections with the exec into separate functions, what about moving everything _before_ the exec into their own functions (maybe `getCommandToCreateFolder` and `getCommandToDelete`)? Then we could still write unit

(Note when moving `ExecCMDInContainer` over from occlient, I couldn't find any references in the existing odo code that had unit tests with it)

Reply:
> (Note when moving `ExecCMDInContainer` over from occlient, I couldn't find any references in the existing odo code that had unit tests for it, or unit tests for functions that called it, so other parts of odo have had this problem too.)

and that is the reason that we should not do this mistake when we have the opportunity to re-write/replace this functionality

Reply:
> As a compromise, rather than putting those sections with the exec into separate functions, what about moving everything _before_ the exec into their own functions (maybe `getCommandToCreateFolder` and `getCommandToDelete`)? Then we could still write unit tests.

I keep that upto you, I just wanted to break this function down into smaller pieces to make it more unit testable

Reply:
@girishramnani I'll go with approach for now (and see if there are any other pieces we can break up). I'll also address your comment regarding the number of parameters.

Since unit tests for the `ExecCmdInContainer` function will require more work / intevestigation / design, I'll address it as part of https://github.com/openshift/odo/issues/2652 (which is also dealing with unit-testing exec)